File: src/utils/info.rs
use std::io::Write;

use tch::nn::VarStore;

pub fn param_count(vs: &VarStore) -> (i64, i64) {
    (
        vs.variables()
            .values()
            .map(|t| t.size().iter().product::<i64>())
            .sum(),
        vs.trainable_variables()
            .iter()
            .map(|t| t.size().iter().product::<i64>())
            .sum(),
    )
}

pub fn write_summary(vs: &VarStore, mut wr: impl Write) -> std::io::Result<()> {
    let mut sorted_vars: Vec<_> = vs.variables().into_iter().collect();
    sorted_vars.sort_by(|a, b| a.0.cmp(&b.0));
    wr.write_fmt(format_args!("{}\n", "-".repeat(85)))?;

    for (name, tensor) in sorted_vars {
        let shape = tensor.size();
        wr.write_fmt(format_args!(
            "{:<60} {:?} {}\n",
            name,
            shape,
            if tensor.requires_grad() {
                "(trainable)"
            } else {
                ""
            }
        ))?;
    }
    wr.write_fmt(format_args!("{}\n", "-".repeat(85)))?;
    Ok(())
}

File: speckle-py/speckle/utils/info.py
import torch


def param_count(model: torch.nn.Module):
    # tch VarStore flags every variable trainable at creation; requires_grad_(false)
    # only freezes autograd, so Rust scheme.txt always reports trainable == total.
    total = sum(t.numel() for t in model.state_dict().values())
    trainable = total
    return total, trainable


def write_summary(model: torch.nn.Module, wr) -> None:
    grad_flags = {name: p.requires_grad for name, p in model.named_parameters()}
    sorted_vars = sorted(model.state_dict().items())
    wr.write("-" * 85 + "\n")
    for name, tensor in sorted_vars:
        shape = str(list(tensor.shape))
        flag = "(trainable)" if grad_flags.get(name, False) else ""
        wr.write(f"{name:<60} {shape} {flag}\n".rstrip() + "\n")
    wr.write("-" * 85 + "\n")
